File: algorithms/algorithms-cplusplus/leetcode/Question_0984_String_Without_AAA_or_BBB.cc
#include <string>

using namespace std;

class Solution {
 public:
  string strWithout3a3b(int a, int b) {
    string res;
    while (a > 0 || b > 0) {
      bool flag = false;
      int l = res.size();
      if (l >= 2 && res[l - 1] == res[l - 2]) {
        if (res[l - 1] == 'b') {
          flag = true;
        }
      } else {
        if (a >= b) {
          flag = true;
        }
      }
      if (flag) {
        a--;
        res.append("a");
      } else {
        b--;
        res.append("b");
      }
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0984_String_Without_AAA_or_BBB.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0984_String_Without_AAA_or_BBB {
    public String strWithout3a3b(int a, int b) {
        StringBuilder res = new StringBuilder();
        while (a > 0 || b > 0) {
            boolean flag = false;
            int l = res.length();
            if (l >= 2 && res.charAt(l - 1) == res.charAt(l - 2)) {
                if (res.charAt(l - 1) == 'b') {
                    flag = true;
                }
            } else {
                if (a >= b) {
                    flag = true;
                }
            }
            if (flag) {
                a--;
                res.append('a');
            } else {
                b--;
                res.append('b');
            }
        }
        return res.toString();
    }
}
